cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -134,195 +160,105 @@ func newEtcdClient() (client *etcd.Client) {
 		}
 		return true, nil
 	}); err != nil {
-		log.Fatalf("Timed out after %s waiting for at least 1 synchronized etcd server in the cluster", timeout)
+		return nil, fmt.Errorf("Timed out after %s waiting for at least 1 synchronized etcd server in the cluster. Error: %v", timeout, err)
 	}
-	return client
+	return client, nil
+}
+
+func getKubeMasterUrl() (string, error) {
+	if *argKubeMasterUrl == "" {
+		return "", fmt.Errorf("no --kube_master_url specified")
+	}
+	parsedUrl, err := url.Parse(os.ExpandEnv(*argKubeMasterUrl))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse --kube_master_url %s - %v", *argKubeMasterUrl, err)
+	}
+	if parsedUrl.Scheme == "" || parsedUrl.Host == "" || parsedUrl.Host == ":" {
+		return "", fmt.Errorf("invalid --kube_master_url specified %s", *argKubeMasterUrl)
+	}
+	return parsedUrl.String(), nil
 }
 
 // TODO: evaluate using pkg/client/clientcmd
 func newKubeClient() (*kclient.Client, error) {
 	var config *kclient.Config
-	if *kubecfg_file == "" {
-		// No kubecfg file provided. Use kubernetes_ro service.
-		masterHost := os.Getenv("KUBERNETES_RO_SERVICE_HOST")
-		if masterHost == "" {
-			log.Fatalf("KUBERNETES_RO_SERVICE_HOST is not defined")
-		}
-		masterPort := os.Getenv("KUBERNETES_RO_SERVICE_PORT")
-		if masterPort == "" {
-			log.Fatalf("KUBERNETES_RO_SERVICE_PORT is not defined")
-		}
+	masterUrl, err := getKubeMasterUrl()
+	if err != nil {
+		return nil, err
+	}
+	if *argKubecfgFile == "" {
 		config = &kclient.Config{
-			Host:    fmt.Sprintf("http://%s:%s", masterHost, masterPort),
-			Version: "v1beta1",
+			Host:    masterUrl,
+			Version: "v1beta3",
 		}
 	} else {
-		masterHost := os.Getenv("KUBERNETES_SERVICE_HOST")
-		if masterHost == "" {
-			log.Fatalf("KUBERNETES_SERVICE_HOST is not defined")
-		}
-		masterPort := os.Getenv("KUBERNETES_SERVICE_PORT")
-		if masterPort == "" {
-			log.Fatalf("KUBERNETES_SERVICE_PORT is not defined")
-		}
-		master := fmt.Sprintf("https://%s:%s", masterHost, masterPort)
 		var err error
 		if config, err = kclientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-			&kclientcmd.ClientConfigLoadingRules{ExplicitPath: *kubecfg_file},
-			&kclientcmd.ConfigOverrides{ClusterInfo: kclientcmdapi.Cluster{Server: master}}).ClientConfig(); err != nil {
+			&kclientcmd.ClientConfigLoadingRules{ExplicitPath: *argKubecfgFile},
+			&kclientcmd.ConfigOverrides{ClusterInfo: kclientcmdapi.Cluster{Server: masterUrl}}).ClientConfig(); err != nil {
 			return nil, err
 		}
 	}
-	log.Printf("Using %s for kubernetes master", config.Host)
-	log.Printf("Using kubernetes API %s", config.Version)
+	glog.Infof("Using %s for kubernetes master", config.Host)
+	glog.Infof("Using kubernetes API %s", config.Version)
 	return kclient.New(config)
 }
 
-func buildNameString(service, namespace, domain string) string {
+func (ks *kube2sky) buildNameString(service, namespace, domain string) string {
```

Comment:
not particular to this line, just general - why make a function into a method if it has no need of the method-receiver?